### PR TITLE
Make sympy.S convert numpy.ndarray to Array

### LINKED ISSUE #10295: sympy.S handles numpy arrays incorrectly.
There are currently (at least) two problems:

First problem is implicit conversion to finite precision floating point values instead of `sympy.Integer` when applicable:

```
In [1]: import numpy, sympy

In [2]: print(numpy.__version__, sympy.__version__)
('1.10.1', '0.7.7.dev')

In [3]: sympy.S(numpy.array([1], dtype=numpy.int64))
Out[3]: 1.00000000000000
```

The other problem is that array data are interpreted as shapes somehow:

```
In [4]: out = sympy.S(numpy.array([1, 2, 3], dtype=numpy.int64))

In [5]: out.shape
Out[5]: (1, 2, 3)

In [6]: out
Out[6]: 
array([[[ 0.,  0.,  0.],
        [ 0.,  0.,  0.]]])
```

This causes major problems when doing e.g. `rref` on a numpy integer array (which we arguably should support).

### PR DESCRIPTION
**Pull request by bjodah**
This commit fixes gh-10295

### CODE REVIEW

**Comment by bjodah** [comment]
Ping @Upabjojr, what do you think?

File: sympy/core/sympify.py [hunk 1]
```diff
@@ -276,6 +276,10 @@ def sympify(a, locals=None, convert_xor=True, strict=False, rational=False,
     if strict:
         raise SympifyError(a)
 
+    if hasattr(a, 'shape'):
```
Comment (Upabjojr) [inline sympy/core/sympify.py]: why not `isinstance(a, numpy.ndarray)` ?

Comment (Upabjojr) [inline sympy/core/sympify.py]: I fear that the `shape` attribute does not identify _numpy_ classes unambiguously. Maybe there are libraries other than _numpy_ defining objects with _shape_ that may be sympified.

**Comment by Upabjojr** [comment]
It's a duplicate of https://github.com/sympy/sympy/pull/10369

I would add a few more tests:
* numpy array with SymPy expression inside it.
* rank-1 import (and check that it's not a matrix).
* rank greater than 2.

Like:
```python
from sympy.abc import x, y
B = numpy.array([-7, x, 3*y**2])
sB = S(B)
```

and 

```python
C = numpy.arange(0, 24)
C.resize(2,3,4)
sC = S(C)
```

with 
```python
>>> C
array([[[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11]],

       [[12, 13, 14, 15],
        [16, 17, 18, 19],
        [20, 21, 22, 23]]])
```

**Comment by Upabjojr** [comment]
Another possible issue, do we want the array to be converted automatically?

In your branch:
```python
>>> x*C
array([[[0, x, 2*x, 3*x],
        [4*x, 5*x, 6*x, 7*x],
        [8*x, 9*x, 10*x, 11*x]],

       [[12*x, 13*x, 14*x, 15*x],
        [16*x, 17*x, 18*x, 19*x],
        [20*x, 21*x, 22*x, 23*x]]], dtype=object)
```

Converting automatically may unfortunately cause serious troubles in _sympy.tensor.tensor_, I would avoid fixing this now.

**Comment by bjodah** [comment]
Ah, I see. I did not think about that case. So perhaps we shouldn't make the choice for the user to use `Array` (maybe they want Matrix etc.).

So perhaps `S` should just sympify the contents of the NumPy array. i.e. we can simply fix the broken `return type(a)([sympify(x, ...` statement to handle NumPy arrays. (which is more in line with what sympify does)

**Comment by Upabjojr** [comment]
Well, I would go for the sympification. Once the array is sympified, SymPy methods can potentially be applied to it (it's still a young module but it can potentially grow up a lot).

`Array` has a `.tomatrix()` method. Matrices are rank-2 arrays only, all other ranks would not be supported. I wouldn't convert directly to a matrix.

> So perhaps S should just sympify the contents of the NumPy array.

I suppose that _S_ is expected to return a SymPy object, because it is often used to make sure that objects provide the _args_ property and SymPy styled hashing.

**Comment by bjodah** [comment]
Yes, returning a SymPy object should be what sympify does. So disregard what I did in that last commit (numpy.vectorize).

So do you think we should map numpy arrays to `Array` but avoid automatic conversion when e.g. multiplying a NumPy array with a Symbol as in your example?

**Comment by Upabjojr** [comment]
> So do you think we should map numpy arrays to Array but avoid automatic conversion when e.g. multiplying a NumPy array with a Symbol as in your example?

I think it's the best way for now.

**Comment by bjodah** [comment]
Alright. Regarding the earlier `hasattr` check, I simply thought checking explicitly for an `numpy.ndarray` was overly restrictive, in the future we might be using e.g. https://github.com/ndarray/ndarray

**Comment by Upabjojr** [comment]
> in the future we might be using e.g. https://github.com/ndarray/ndarray

Looks like a nice library for _symengine_.

**Comment by Upabjojr** [comment]
Reopened as `sympy.tensor.tensor` does not depend on `numpy` anymore.

**Comment by Upabjojr** [comment]
OK, I cloned this PR.

**Pull request closed without merging**